Remove all references to 'grained' access control, and reword a few t…

mostekcm opened the pull request:
…hings
 
Tami, really want Peter's review before we merge this one.  Had to reword some stuff based on a conversation with architecture.

mostekcm committed:
Remove all references to 'grained' access control, and reword a few things

yvonnewilson commented:
The advice about OAuth being overkill for "private" APIs (part of app) - did that already get reviewed? I sensed hesitancy on saying that publicly, at least from Jared earlier.

the user's data (the reseource server) <- "resource" is not spelled correctly there.

PeterGFernandez committed:
Update _introduction.md

Re-worded the introductory paragraph in line with the re-factor of the terminology.

PeterGFernandez committed:
Update _introduction.md

Re-factored to reference "categories" rather than "levels". This fits better within the context of the guidance provides in that (a) there is not implication of hierarchy (as there typically is with the term "level") and (b) the term category is already used in this context elsewhere in the narrative.

PeterGFernandez committed:
Update _introduction.md

Re-factored grammar for better understanding

PeterGFernandez committed:
Update _introduction.md

Re-factored to remove ambiguity.

PeterGFernandez committed:
Update _introduction.md

Re-factored to remove ambiguity

PeterGFernandez reviewed (dismissed):
I've made a number of suggested re-factors to improve clarity, and also make the terminology less "controversial" from a customers perspective

tamigoodall committed:
Small grammar fixes

Small grammar fixes

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,38 @@
-It’s important to start by distinguishing between Authentication, Authorization, and Access Control. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and some or all of Authorization. Access Control however must be the responsibility of the Application or API itself, because access control is almost always contextual: 
+It’s important to start by distinguishing between Authentication, Authorization, and Access Control. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and some or all of Authorization and Access Control. Additionally, ab Application or API itself may need to share Access Control responsibility, especially where contextual access is required: 
```
```suggestion
It’s important to start by distinguishing between Authentication, Authorization, and Access Control. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and some or all of Authorization and Access Control. Additionally, an Application or API itself may need to share Access Control responsibility, especially where contextual access is required: 
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,38 @@
-  * **Access Control**: the access that is permitted. In general this can be thought of as breaking down into three sub-categories of granularity: 
+* **Authorization**: the proceess of determining what is allowed, based on the principal, the consent they provide, what permissions they have been given, and/or the set of contextually specific access criteria.
+* **Consent**: what can be done on the behalf of a user principal (subject) or, in the case of an application principal authenticating via [Client Credentials](/flows/concepts/client-credentials) grant, what an administrative authority has authorized an application to do.
+* **Access Control**: the access that is permitted. In general we typically group different types of access control into three distinct categories, so that it's easier to understand (a) which actor is responsible for storing the information, (b) which actor is responsible for making decisions, and (c) which is responsible for enforcing the restrictions. 
```
```suggestion
* **Access Control**: Access control is really the umbrella concept that puts Authentication, Authorization, and Enforcement into one bucket.  In general we typically group different types of access control into three distinct categories, so that it's easier to understand (a) which actor is responsible for storing the information, (b) which actor is responsible for making decisions, and (c) which is responsible for enforcing the restrictions. 
```

mostekcm committed:
Update articles/architecture-scenarios/_includes/_authorization/_introduction.md

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,38 @@
+* **Authentication**: the process of determining if a principal (a user or application) is who or what they say they are.
+* **Authorization**: the proceess of determining what is allowed, based on the principal, the consent they provide, what permissions they have been given, and/or the set of contextually specific access criteria.
+* **Consent**: what can be done on the behalf of a user principal (subject) or, in the case of an application principal authenticating via [Client Credentials](/flows/concepts/client-credentials) grant, what an administrative authority has authorized an application to do.
+* **Access Control**: the access that is permitted. In general we typically group different types of access control into three distinct categories so that it's easier to understand a) which actor is responsible for storing the information, b) which actor is responsible for making decisions, and c) which is responsible for enforcing the restrictions. 
```
```suggestion
* **Access Control**: This is an umbrella concept that ties together Authentication, Authorization, and Enforcement of policies.  In general we typically group different types of access control into three distinct categories so that it's easier to understand a) which actor is responsible for storing the information, b) which actor is responsible for making decisions, and c) which is responsible for enforcing the restrictions. 
```

mostekcm committed:
Update articles/architecture-scenarios/_includes/_authorization/_introduction.md

mostekcm committed:
Another refactor to give Access Control a more broad definition and introduce the concept of Policy Enforcement

iamJpRowan commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
 * Will your APIs also be accessed by your own (first-party) applications?
 * Will your application be calling a third-party API?
-* Should your applications and/or APIs be enforcing access control based on medium or fine-grained permissions?
+* Should your applications and/or APIs be enforcing access control based on user information?
```
user "claims" - to align with SAML and OIDC terminology?

mostekcm replied:
```suggestion
* Should your applications and/or APIs be enforcing access control based on user claims?
```

iamJpRowan commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
-It’s important to start by distinguishing between Authentication, Authorization, and Access Control. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and some or all of Authorization. Access Control however must be the responsibility of the Application or API itself, because access control is almost always contextual: 
+Let's start about by taking a step back and talking about Access Control.  There isn't one clear cut definition of Access Control in the industry, but if you spend some time searching and reading you'll see that most authoritative sources agree that it is the umbrella concept that puts all of Authentication, Authorization, Consent, and Policy Enforcment together to ensure that only the right peop ...
```
misspelled enforcement in the first instance here

mostekcm replied:
```suggestion
Let's start by taking a step back and talking about Access Control.  There isn't one clear cut definition of Access Control in the industry, but if you spend some time searching and reading you'll see that most authoritative sources agree that it is the umbrella concept that puts all of Authentication, Authorization, Consent, and Policy Enforcement together to ensure that only the right people and services have access to your applications and APIs.  Next, let's dive more closely into the distinctions between Authentication, Authorization, Consent, and Policy Enforcement. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and Consent, and some or all of Authorization and Policy Enforcement. Additionally, an Application or API itself almost always is the primary enforcer of policies, especially where contextual access is required: 
```

mostekcm committed:
Update articles/architecture-scenarios/_includes/_authorization/_introduction.md

PeterGFernandez commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
 
 ::: warning
-When deciding what data to include in OIDC tokens, consider token size, especially if you are passing the token in the URL. Even if you are not passing tokens in the URL, you will also need to consider the potential of exposing sensitive PII (Personally Identifiable Information). Token information is not encrypted, so although it isn't generally a security issue for an ID token to be leaked, it c ...
+When deciding what data to include in your ID token and/or access token, consider token size, especially if you are passing the token in the URL. Even if you are not passing tokens in the URL, you will also need to consider the potential of exposing sensitive PII (Personally Identifiable Information). Token information is not encrypted, so although it isn't generally a security issue for an ID to ...
```
Back-to-back warning panels feels a little "clunky" from a UX perspective.

mostekcm replied:
removing first warning in lieu of inline comments.

PeterGFernandez commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
-It’s important to start by distinguishing between Authentication, Authorization, and Access Control. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and some or all of Authorization. Access Control however must be the responsibility of the Application or API itself, because access control is almost always contextual: 
+Let's start by taking a step back and talking about Access Control.  There isn't one clear cut definition of Access Control in the industry, but if you spend some time searching and reading you'll see that most authoritative sources agree that it is the umbrella concept that puts all of Authentication, Authorization, Consent, and Policy Enforcement together to ensure that only the right people an ...
```
Does one "dive more closely"? I'd say perhaps one would "dive deeper" or "look more closely"?

mostekcm replied:
```suggestion
Let's start by taking a step back and talking about Access Control.  There isn't one clear cut definition of Access Control in the industry, but if you spend some time searching and reading you'll see that most authoritative sources agree that it is the umbrella concept that puts all of Authentication, Authorization, Consent, and Policy Enforcement together to ensure that only the right people and services have access to your applications and APIs.  Next, let's look more closely into the distinctions between Authentication, Authorization, Consent, and Policy Enforcement. Your Auth0 tenant (your Authorization Server) is typically responsible for Authentication and Consent, and some or all of Authorization and Policy Enforcement. Additionally, an Application or API itself almost always is the primary enforcer of policies, especially where contextual access is required: 
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
 :::
 
-Auth0 also can provide the authorization information needed so that the application or API can apply medium or fine-grained access control. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decid ...
+Auth0 also can provide the information needed so that an application or API can enforce restrictions. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decide what information you require for you ...
```
```suggestion
Auth0 also can provide the information needed so that an application can enforce restrictions. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to perform policy enforcement. In this case you will need to decide what information you require for your application to make enforcement decisions.  If you need to make decisions at an API instead of in your application, you will likely need to use an Access Token instead of an ID token.  Continue reading for more information.
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
 
-Auth0 also can provide the authorization information needed so that the application or API can apply medium or fine-grained access control. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decid ...
+Auth0 also can provide the information needed so that an application or API can enforce restrictions. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decide what information you require for you ...
+
```
```suggestion
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
-Auth0 also can provide the authorization information needed so that the application or API can apply medium or fine-grained access control. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decid ...
+Auth0 also can provide the information needed so that an application or API can enforce restrictions. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decide what information you require for you ...
+
+::: warning
```
```suggestion
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
+Auth0 also can provide the information needed so that an application or API can enforce restrictions. For [application level integration](#application-integration), Auth0 allows you to add [custom claims](#id-token-claims) to an ID Token, which your application can then verify and subsequently use to enforce access control. In this case you will need to decide what information you require for you ...
+
+::: warning
+ID Tokens should only be used within the application itself, it should not be used as a Bearer token to call an API.  If you need a token to call an API you should request an Access token instead.  See [API level integration](#api-integration) for more information.
```
```suggestion
```

mostekcm commented on articles/architecture-scenarios/_includes/_authorization/_introduction.md:
```diff
@@ -1,34 +1,44 @@
+
+::: warning
+ID Tokens should only be used within the application itself, it should not be used as a Bearer token to call an API.  If you need a token to call an API you should request an Access token instead.  See [API level integration](#api-integration) for more information.
+:::
```
```suggestion
```

mostekcm committed:
Apply suggestions from code review

Make some minor adjustments based on code review.

yvonnewilson commented:
My $.02 worth:
in _api-integration.md:
 - it says "typically expressed as a JWT". Access tokens in original OAuth 2.0 spec were not JWTs, so this statement might need some qualification such as "with Auth0" or "with Auth0 for custom APIs"?

in introduction.md:
"some or all of Authorization and Policy Enforcement" - I don't think Auth0 can do 100% of policy enforcement in many cases. It can certainly do some, with rules denying access in some cases, and it can probably do more for consumer-facing cases than the enterprise-facing cases, but in any case where rules don't stop the transaction, for example, an API still has to do some of the enforcement.

Authorization the process of determining what is allowed - "specifying" may be better than "determining".

"The user has to give permission to the Client (Relying Party) to access the user's data in a different system."  Vittorio just pointed out to me elsewhere that in OAuth 2.0 it is the API that is the relying party, not the client application.

In the section on the second category, "The data required to enforce this is typically stored in the Authorization Server" - is it worth mentioning that it could be obtained/refreshed from a connection or possibly retrieved by a rule from a policy server (and possibly explaining why the latter might be a problem)? e.g. while it is stored in the authorization server, that might not be the system of record. The following "F" in "For example..." should not be capitalized.

In the third category - the data could be stored in authorization server if it is another attribute. 

"Even if you are not passing tokens in the URL" - it might be helpful to mention what the alternative option is.

"For [API level integration]..." - I expected to see "authorization" or "enforcement" here instead of "integration"

"scope ...reconfiguration" - has this term been defined?

tamigoodall committed:
Merge branch 'master' into mostekcm-fix-authorization-based-on-arch-review
